src/utils/daos_hdlr.c

```diff
@@ -1566,11 +1598,1197 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+{
+	struct dirent* entry = NULL;
+
+	if (file_dfs->type == POSIX) {
+		entry = readdir(dirp);
+	} else if (file_dfs->type == DAOS) {
+		entry = readdir_dfs(file_dfs, dirp);
+	} else {
+		fprintf(stderr, "File type not known, type=%d\n",
+			file_dfs->type);
+	}
+	return entry;
+}
+
+static int
+stat_dfs(struct file_dfs *file_dfs, const char *path, struct stat *buf)
+{
+	int 		rc	  = 0;
+	dfs_obj_t 	*parent	  = NULL;
+	char 		*name	  = NULL;
+	char 		*dir_name = NULL;
+
+	parse_filename_dfs(path, &name, &dir_name);
+	assert(dir_name);
+	/* Lookup the parent directory */
+	rc = dfs_lookup(file_dfs->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+	if (parent == NULL) {
+		fprintf(stderr, "dfs_lookup %s failed, %d \n", dir_name, rc);
+		errno = rc;
+		D_GOTO(out, rc);
+	} else {
+		/* Stat the path */
+		rc = dfs_stat(file_dfs->dfs, parent, name, buf);
+		if (rc) {
+			fprintf(stderr, "dfs_stat %s failed (%d %s)\n",
+				name, rc, strerror(rc));
+			errno = rc;
+		}
+	}
+out:
+	if (parent != NULL) {
+		rc = dfs_release(parent);
+		if (rc != 0) {
+			fprintf(stderr, "dfs_release %s failed with error %d\n",
+				dir_name, rc);
+		}
+	}
+	if (name != NULL)
+		free(name);
+	if (dir_name != NULL)
+		free(dir_name);
+	return rc;
+}
+
+static int
+file_lstat(struct file_dfs *file_dfs, const char *path, struct stat *buf)
+{
+	int rc = 0;
+
+	if (file_dfs->type == POSIX) {
+		rc = lstat(path, buf);
+	} else if (file_dfs->type == DAOS) {
+		rc = stat_dfs(file_dfs, path, buf);
+	} else {
+		fprintf(stderr, "File type not known, file=%s, type=%d\n",
+			path, file_dfs->type);
+	}
+	return rc;
+}
+
+static ssize_t
+read_dfs(struct file_dfs *file_dfs,
+	 const char *file,
+	 void *buf,
+	 size_t size)
+{
+	d_iov_t		iov;
+	d_sg_list_t	sgl;
+	daos_size_t	got_size;
+
+	d_iov_set(&iov, buf, size);
+	sgl.sg_nr = 1;
+	sgl.sg_iovs = &iov;
+	sgl.sg_nr_out = 1;
+
+	/* execute read operation */
+	d_iov_set(&iov, buf, size);
+	int rc = dfs_read(file_dfs->dfs, file_dfs->obj, &sgl,
+			  file_dfs->offset, &got_size, NULL); 
+	if (rc) {
+		fprintf(stderr, "dfs_read %s failed (%d %s)\n",
+			file, rc, strerror(rc));
+		errno = rc;
+		D_GOTO(out, rc);
+	}
+
+	/* update file pointer with number of bytes read */
+	file_dfs->offset += (daos_off_t)got_size;
+out:
+	return (ssize_t)got_size;
+}
+
+static ssize_t
+file_read(struct file_dfs *file_dfs, char *file,
+		void *buf, size_t size)
+{
+	ssize_t got_size = 0;
+
+	if (file_dfs->type == POSIX) {
+		got_size = read(file_dfs->fd, buf, size);
+	} else if (file_dfs->type == DAOS) {
+		got_size = read_dfs(file_dfs, file, buf, size);
+	} else {
+		got_size = 1;
+		fprintf(stderr, "File type not known: %s type=%d\n",
+			file, file_dfs->type);
+	}
+	return got_size;
+}
+
+static int
+closedir_dfs(DIR *_dirp)
+{
+	struct	dfs_daos *dirp	= (struct dfs_daos *)_dirp;
+	int	rc		= dfs_release(dirp->dir);
+
+	if (rc) {
+		fprintf(stderr, "dfs_release failed (%d %s)\n",
+			rc, strerror(rc));
+		errno = rc;
+		rc = 1;
```

Comment:
@daltonbohning yeah, since it was mentioned I was returning both negative and positive error codes I went with positive. Since the dfs_ functions all return positive error codes. What I thought I would do is check if there was an error on a POSIX function, then set the rc to the errno since the POSIX functions do set that, and its more consistent with what the dfs_ functions return?

Reply:
@dsikich Returning positive errno codes is more consistent with the DFS API, but not with the POSIX API.
Most of the POSIX syscalls return -1 on error and set errno accordingly.
So if we want these DFS wrapper functions to be consistent with the POSIX equivalents, I think we should return -1 on error and set errno accordingly.
Thoughts?

Reply:
for wrappers i prefer to match what DFS does for the error code. it's usually tricky to mess with errno especially at a user space level. but i don't have a strong opinion here. i remember once before Ashley did, so maybe discuss with him?

Reply:
@mchaarawi I was thinking maybe we could cross that bridge as part of [DAOS-6098](https://jira.hpdd.intel.com/browse/DAOS-6098). I.e. as long as error checks are in place here, we can worry about setting errno vs returning an errno later, which will be discussed as part of [DAOS-6098](https://jira.hpdd.intel.com/browse/DAOS-6098).
Are you okay with this?

Reply:
yes but in this case then we should either go back to return -1 or the actual error code.
1 is weird to me and wrong in the case of read.

Reply:
@mchaarawi yeah returning 1 for bytes_read is weird I can update that.

Reply:
Yes, I agree